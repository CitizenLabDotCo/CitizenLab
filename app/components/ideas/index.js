import React from 'react';
import PropTypes from 'prop-types';

// components
import InfiniteScroll from 'react-infinite-scroller';
import IdeaCard from 'components/IdeaCard';


// store
import { preprocess } from 'utils';
import { createStructuredSelector } from 'reselect';
import { loadIdeasRequest } from 'resources/ideas/actions';

const IdeasCards = ({ ideas, hasMore, loadMoreIdeas }) => (
  <InfiniteScroll
    element={'div'}
    loadMore={loadMoreIdeas}
    className={'ui stackable cards'}
    initialLoad
    hasMore={hasMore}
    loader={<div className="loader"></div>}
  >
    {ideas.map((id) => (
      <IdeaCard key={id} id={id} />
    ))}
  </InfiniteScroll>
);

IdeasCards.propTypes = {
  ideas: PropTypes.any,
  hasMore: PropTypes.bool,
  loadMoreIdeas: PropTypes.func.isRequired,
};

const mapStateToProps = createStructuredSelector({
  ideas: (state) => state.getIn(['ideasIndex', 'list']),
  nextPageNumber: (state) => state.getIn(['ideasIndex', 'nextPageNumber']),
  nextPageItemCount: (state) => state.getIn(['ideasIndex', 'nextPageItemCount']),
  search: (state) => state.getIn(['route', 'locationBeforeTransitions', 'search']),
});

const mergeProps = (state, dispatch, own) => {
  const { ideas, nextPageNumber, nextPageItemCount, search } = state;
  const { load } = dispatch;
  const { filter, maxNumber } = own;
  let { hasMore } = own;
  if (hasMore !== false) hasMore = !!(nextPageNumber && (maxNumber || nextPageItemCount));

  return {
    loadMoreIdeas: () => load(nextPageNumber, maxNumber || nextPageItemCount, search, filter),
    hasMore,
    ideas,
  };
};


export default preprocess(mapStateToProps, { load: loadIdeasRequest }, mergeProps)(IdeasCards);
