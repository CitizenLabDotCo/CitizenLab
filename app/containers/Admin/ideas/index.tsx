import React from 'react';

// router
import { withRouter, WithRouterProps } from 'react-router';

// components
import HelmetIntl from 'components/HelmetIntl';
import TabbedResource, { TabProps } from 'components/admin/TabbedResource';

// i18n
import messages from './messages';
import { InjectedIntlProps } from 'react-intl';
import { injectIntl } from 'utils/cl-intl';

export interface InputProps {}

interface DataProps {
  location;
}

interface Props extends InputProps, DataProps {}

interface State {}

class IdeasPage extends React.PureComponent<
  Props & InjectedIntlProps & WithRouterProps,
  State
> {
  getTabs = () => {
    const { formatMessage } = this.props.intl;

    let tabs: TabProps[] = [
      {
        label: formatMessage(messages.tabSettings),
        url: '/admin/ideas',
      },
      {
        label: formatMessage(messages.tabCustomize),
        url: '/admin/ideas/statuses',
        active: this.props.location.pathname.includes('/admin/ideas/statuses'),
      },
    ];
    return tabs;
  };

  render() {
    const { children } = this.props;
    const { formatMessage } = this.props.intl;

    const resource = {
      title: formatMessage(messages.pageTitle),
    };

    return (
      <TabbedResource resource={resource} tabs={this.getTabs()}>
        <HelmetIntl
          title={messages.helmetTitle}
          description={messages.helmetDescription}
        />
        {children}
      </TabbedResource>
    );
  }
}

const IdeasPageWithHocs = withRouter(injectIntl(IdeasPage));

export default (inputProps: InputProps & WithRouterProps) => (
  <IdeasPageWithHocs {...inputProps} />
);
