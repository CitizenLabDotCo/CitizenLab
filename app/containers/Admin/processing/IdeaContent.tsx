import React, { PureComponent } from 'react';
import { isNilOrError, getFormattedBudget } from 'utils/helperUtils';
import { adopt } from 'react-adopt';

// components
import Title from 'components/PostShowComponents/Title';
import Body from 'components/PostShowComponents/Body';
import IdeaProposedBudget from 'containers/IdeasShow/IdeaProposedBudget';
import { Tag } from 'cl2-component-library';

// resources
import GetIdea, { GetIdeaChildProps } from 'resources/GetIdea';
import GetTenant, { GetTenantChildProps } from 'resources/GetTenant';
import GetLocale, { GetLocaleChildProps } from 'resources/GetLocale';

// i18n
import injectLocalize, { InjectedLocalized } from 'utils/localize';
import { injectIntl, FormattedMessage } from 'utils/cl-intl';
import { InjectedIntlProps } from 'react-intl';
import messages from './messages';

// style
import styled from 'styled-components';
import { stylingConsts } from 'utils/styleUtils';

const Content = styled.div`
  width: 100%;
  flex: 5;
  height: 100%;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
`;

const StyledBody = styled(Body)`
  min-height: calc(100vh - ${stylingConsts.menuHeight}px - 200px);
  overflow-y: auto;
`;

const TagList = styled.div`
  text-align: right;
  width: 100%;
  margin-bottom: 10px;
`;

const StyledTitle = styled(Title)`
  margin-bottom: 20px;
`;

const BodySectionTitle = styled.h2`
  font-size: ${(props) => props.theme.fontSizes.medium}px;
  font-weight: 400;
  line-height: 28px;
`;

export interface InputProps {
  ideaId: string | null;
}

interface DataProps {
  idea: GetIdeaChildProps;
  tenant: GetTenantChildProps;
  locale: GetLocaleChildProps;
}

interface Props extends InputProps, DataProps {}

export class IdeaContent extends PureComponent<
  Props & InjectedLocalized & InjectedIntlProps
> {
  constructor(props) {
    super(props);
  }

  render() {
    const { idea, localize, tenant, locale } = this.props;

    if (!isNilOrError(idea) && !isNilOrError(locale) && !isNilOrError(tenant)) {
      const ideaId = idea.id;
      const ideaTitle = localize(idea.attributes.title_multiloc);
      // AuthorId can be null if user has been deleted
      const proposedBudget = idea.attributes.proposed_budget;
      const currency = tenant.attributes.settings.core.currency;

      return (
        <Content>
          <StyledTitle postId={ideaId} postType="idea" title={ideaTitle} />
          {proposedBudget && (
            <>
              <BodySectionTitle>
                <FormattedMessage {...messages.proposedBudgetTitle} />
              </BodySectionTitle>
              <IdeaProposedBudget
                formattedBudget={getFormattedBudget(
                  locale,
                  proposedBudget,
                  currency
                )}
              />
              <BodySectionTitle>
                <FormattedMessage {...messages.description} />
              </BodySectionTitle>
            </>
          )}

          <TagList>
            <Tag isAutoTag={false} isSelected={false} text="tag one" />
            <Tag isAutoTag={false} isSelected={false} text="tag one" />
            <Tag isAutoTag={false} isSelected={false} text="tag one" />
            <Tag isAutoTag={false} isSelected={false} text="tag one" />
          </TagList>
          <StyledBody
            postId={ideaId}
            postType="idea"
            body={localize(idea.attributes.body_multiloc)}
            locale={locale}
          />
        </Content>
      );
    }
    return null;
  }
}

const Data = adopt<DataProps, InputProps>({
  tenant: <GetTenant />,
  locale: <GetLocale />,
  idea: ({ ideaId, render }) => <GetIdea ideaId={ideaId}>{render}</GetIdea>,
});

const IdeaContentWithHOCs = injectIntl(injectLocalize(IdeaContent));

const WrappedIdeaContent = (inputProps: InputProps) => (
  <Data {...inputProps}>
    {(dataProps) => <IdeaContentWithHOCs {...inputProps} {...dataProps} />}
  </Data>
);

export default WrappedIdeaContent;
