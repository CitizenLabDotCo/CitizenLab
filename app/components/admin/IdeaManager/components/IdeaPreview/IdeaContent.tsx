import React, { PureComponent } from 'react';
import { isNilOrError } from 'utils/helperUtils';
import { adopt } from 'react-adopt';

// components
import IdeaAuthor from 'containers/IdeasShow/IdeaAuthor';
import IdeaHeader from 'containers/IdeasShow/IdeaHeader';
import OfficialFeedback from 'containers/IdeasShow/OfficialFeedback';
import Comments from 'containers/IdeasShow/Comments';
import FileAttachments from 'components/UI/FileAttachments';
import Icon from 'components/UI/Icon';
import IdeaSettings from './IdeaSettings';
import VotePreview from './VotePreview';
import InfoTooltip from 'components/admin/InfoTooltip';
import Button from 'components/UI/Button';
import { Top, Content, Container } from '.';

// resources
import IdeaBody from 'containers/IdeasShow/IdeaBody';
import IdeaMap from 'containers/IdeasShow/IdeaMap';
import GetResourceFiles, { GetResourceFilesChildProps } from 'resources/GetResourceFiles';
import GetIdea, { GetIdeaChildProps } from 'resources/GetIdea';
import GetIdeaImages, { GetIdeaImagesChildProps } from 'resources/GetIdeaImages';
import GetTenant, { GetTenantChildProps } from 'resources/GetTenant';
import { deleteIdea } from 'services/ideas';

// i18n
import injectLocalize, { InjectedLocalized } from 'utils/localize';
import { injectIntl, FormattedMessage } from 'utils/cl-intl';
import { InjectedIntlProps, FormattedNumber } from 'react-intl';
import messages from './messages';

// animations
import CSSTransition from 'react-transition-group/CSSTransition';

// style
import styled from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';
import { darken } from 'polished';
import { get } from 'lodash-es';

const Row = styled.div`
  display: flex;
  width: 100%;
`;

const Left = styled.div`
  flex: 5;
  margin-right: 50px;
  height: 100%;
`;

const IdeaImage = styled.img`
  width: 100%;
  margin: 0 0 2rem;
  padding: 0;
  border-radius: 8px;
  border: 1px solid ${colors.separation};
`;

const MapWrapper = styled.div`
  border-radius: 8px;
  border: 1px solid ${colors.separation};
  height: 265px;
  position: relative;
  overflow: hidden;
  z-index: 2;

  &.map-enter {
    height: 0;
    opacity: 0;

    &.map-enter-active {
      height: 265px;
      opacity: 1;
      transition: all 250ms ease-out;
    }
  }

  &.map-exit {
    height: 265px;
    opacity: 1;

    &.map-exit-active {
      height: 0;
      opacity: 0;
      transition: all 250ms ease-out;
    }
  }
`;

const AddressWrapper = styled.div`
  color: #fff;
  font-size: ${fontSizes.base}px;
  font-weight: 300;
  background: rgba(0, 0, 0, 0.4);
  border-top: 1px solid ${colors.separation};
  margin: 0;
  padding: 10px;
  position: absolute;
  right: 0;
  left: 0;
  bottom: 0;
  z-index: 1000;
`;

const StyledOfficialFeedback = styled(OfficialFeedback)`
  margin-top: 70px;
`;

const Right = styled.div`
  flex: 2;
  position: sticky;
  top: 80px;
  align-self: flex-start;
  color: ${colors.adminTextColor};
  font-size: ${fontSizes.base}px;
  line-height: 19px;
`;

const BudgetBox = styled.div`
  margin-top: 20px;
  width: 100%;
  height: 95px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 5px;
  position: relative;
  border-radius: 5px;
  background: ${colors.background};
  border: solid 1px ${colors.adminTextColor};
  font-size: ${fontSizes.large}px;
  font-weight: 500;
`;

const Picks = styled.div`
  margin-top: 7px;
  display: flex;
  font-size: ${fontSizes.base}px;
  align-items: center;
`;

const LocationIconWrapper = styled.div`
  width: 22px;
  height: 36px;
  margin: 0;
  margin-right: 3px;
  padding: 0;
  border: none;
  display: flex;
  align-items: center;
  justify-content: flex-start;
`;

const LocationIcon = styled(Icon)`
  width: 18px;
  fill: ${colors.adminTextColor};
`;

const LocationButton = styled.button`
  display: flex;
  align-items: center;
  margin-bottom: 30px;
  margin-top: 20px;
  margin-right: 6px;
  text-align: left;
  font-weight: 400;
  transition: all 100ms ease-out;
  white-space: nowrap;

  &:hover {
    color: ${darken(0.2, colors.adminTextColor)};

    ${LocationIcon} {
      fill: ${darken(0.2, colors.adminTextColor)};
    }
  }
`;

interface State {
  showMap: boolean;
}

interface InputProps {
  ideaId: string | null;
  closeSideModal: () => void;
  handleClickEdit: () => void;
}

interface DataProps {
  idea: GetIdeaChildProps;
  ideaImages: GetIdeaImagesChildProps;
  ideaFiles: GetResourceFilesChildProps;
  tenant: GetTenantChildProps;
}

interface Props extends InputProps, DataProps {}

export class IdeaContent extends PureComponent<Props & InjectedLocalized & InjectedIntlProps, State> {
  constructor(props) {
    super(props);
    this.state = {
      showMap: false
    };
  }

  handleClickDelete = () => {
    const { idea, closeSideModal } = this.props;
    const message = this.props.intl.formatMessage(messages.deleteIdeaConfirmation);

    if (!isNilOrError(idea)) {
      if (window.confirm(message)) {
        deleteIdea(idea.id);
        closeSideModal();
      }
    }
  }

  handleMapWrapperSetRef = (element: HTMLDivElement) => {
    if (element) {
      element.scrollIntoView({
        behavior: 'smooth',
        block: 'center',
        inline: 'nearest'
      });
    }
  }

  handleMapToggle = () => {
    this.setState((state) => {
      const showMap = !state.showMap;
      return { showMap };
    });
  }

  render() {
    const { idea, localize, ideaImages, ideaFiles, tenant, handleClickEdit, intl: { formatMessage } } = this.props;
    const { showMap } = this.state;
    if (!isNilOrError(idea)) {
      const ideaTitle = localize(idea.attributes.title_multiloc);

      const ideaImageLarge = !isNilOrError(ideaImages) && ideaImages.length > 0 ? get(ideaImages[0], 'attributes.versions.large', null) : null;

      const ideaLocation = (idea.attributes.location_point_geojson || null);
      const ideaAdress = (idea.attributes.location_description || null);

      return (
        <Container>
          <Top>
            <Button
              icon="edit"
              style="text"
              textColor={colors.adminTextColor}
              onClick={handleClickEdit}
            >
              <FormattedMessage {...messages.edit}/>
            </Button>
            <Button
              icon="delete"
              style="text"
              textColor={colors.adminTextColor}
              onClick={this.handleClickDelete}
            >
              <FormattedMessage {...messages.delete}/>
            </Button>
          </Top>
          <Content>
            <IdeaHeader
              ideaId={idea.id}
              ideaTitle={ideaTitle}
              projectId={idea.relationships.project.data.id}
            />
            <Row>
              <Left>
                {ideaImageLarge &&
                  <IdeaImage src={ideaImageLarge} alt={formatMessage(messages.imageAltText, { ideaTitle })} className="e2e-ideaImage"/>
                }
                <IdeaAuthor
                  authorId={get(idea, 'relationships.author.data.id', null)}
                  ideaCreatedAt={idea.attributes.created_at}
                  ideaId={idea.id}
                />
                <IdeaBody
                  ideaId={idea.id}
                  ideaBody={localize(idea.attributes.body_multiloc)}
                />
                {ideaLocation &&
                  <>
                    <CSSTransition
                      classNames="map"
                      in={showMap}
                      timeout={300}
                      mountOnEnter={true}
                      unmountOnExit={true}
                      exit={true}
                    >
                      <MapWrapper innerRef={this.handleMapWrapperSetRef}>
                        <IdeaMap location={ideaLocation} id={idea.id} />
                        {ideaAdress && <AddressWrapper>{ideaAdress}</AddressWrapper>}
                      </MapWrapper>
                    </CSSTransition>
                  </>
                }

                {ideaFiles && !isNilOrError(ideaFiles) &&
                  <FileAttachments files={ideaFiles} />
                }

                <StyledOfficialFeedback
                  ideaId={idea.id}
                />

                <Comments ideaId={idea.id} />
              </Left>
              <Right>
                <VotePreview ideaId={idea.id}/>

                {idea.attributes.budget && !isNilOrError(tenant) &&
                  <>
                    <BudgetBox>
                      <FormattedNumber
                        value={idea.attributes.budget}
                        style="currency"
                        currency={tenant.attributes.settings.core.currency}
                        minimumFractionDigits={0}
                        maximumFractionDigits={0}
                      />
                      <Picks>
                        <FormattedMessage {...messages.picks} values={{ picksNumber: idea.attributes.baskets_count }} />
                        &nbsp;
                        <InfoTooltip {...messages.basketsCountTooltip} size="small" position="top-left" />
                      </Picks>
                    </BudgetBox>
                  </>
                }

                <IdeaSettings ideaId={idea.id}/>

                {ideaLocation &&
                  <LocationButton onClick={this.handleMapToggle}>
                    <LocationIconWrapper>
                      <LocationIcon name={!showMap ? 'position' : 'close'} />
                    </LocationIconWrapper>
                      {!showMap
                        ? <FormattedMessage {...messages.openMap} />
                        : <FormattedMessage {...messages.closeMap} />
                      }
                  </LocationButton>
                }
              </Right>
            </Row>
          </Content>
        </Container>
      );
    }
    return null;
  }
}

const Data = adopt<DataProps, InputProps>({
  idea: ({ ideaId, render }) => <GetIdea id={ideaId}>{render}</GetIdea>,
  ideaFiles: ({ ideaId, render }) => <GetResourceFiles resourceId={ideaId} resourceType="idea">{render}</GetResourceFiles>,
  ideaImages: ({ ideaId, render }) => <GetIdeaImages ideaId={ideaId}>{render}</GetIdeaImages>,
  tenant: <GetTenant />
});

const IdeaContentWithHOCs = injectIntl(injectLocalize(IdeaContent));

const WrappedIdeaContent = (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <IdeaContentWithHOCs {...inputProps} {...dataProps} />}
  </Data>
);

export default WrappedIdeaContent;
