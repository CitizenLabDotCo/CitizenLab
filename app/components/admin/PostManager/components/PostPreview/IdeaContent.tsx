import React, { PureComponent } from 'react';
import { isNilOrError } from 'utils/helperUtils';
import { adopt } from 'react-adopt';
import { get } from 'lodash-es';

// components
import IdeaAuthor from 'containers/IdeasShow/IdeaAuthor';
import IdeaTitle from 'containers/IdeasShow/IdeaTitle';
import IdeaBody from 'containers/IdeasShow/IdeaBody';
import IdeaMap from 'containers/IdeasShow/IdeaMap';
import OfficialFeedback from 'containers/IdeasShow/OfficialFeedback';
import Comments from 'containers/IdeasShow/Comments';
import FileAttachments from 'components/UI/FileAttachments';
import IdeaSettings from './IdeaSettings';
import VotePreview from './VotePreview';
import InfoTooltip from 'components/admin/InfoTooltip';
import Button from 'components/UI/Button';
import Link from 'utils/cl-router/Link';
import T from 'components/T';
import { Top, Content, Container } from '.';

// srvices
import { deleteIdea } from 'services/ideas';

// resources
import GetResourceFiles, { GetResourceFilesChildProps } from 'resources/GetResourceFiles';
import GetIdea, { GetIdeaChildProps } from 'resources/GetIdea';
import GetIdeaImages, { GetIdeaImagesChildProps } from 'resources/GetIdeaImages';
import GetTenant, { GetTenantChildProps } from 'resources/GetTenant';
import GetProject, { GetProjectChildProps } from 'resources/GetProject';

// i18n
import injectLocalize, { InjectedLocalized } from 'utils/localize';
import { injectIntl, FormattedMessage } from 'utils/cl-intl';
import { InjectedIntlProps, FormattedNumber } from 'react-intl';
import messages from './messages';

// style
import styled from 'styled-components';
import { colors, fontSizes } from 'utils/styleUtils';
import { darken } from 'polished';

const StyledIdeaTitle = styled(IdeaTitle)`
  margin-bottom: 30px;
`;

const Row = styled.div`
  display: flex;
  width: 100%;
`;

const Left = styled.div`
  flex: 5;
  margin-right: 50px;
  height: 100%;
`;

const BelongsToProject = styled.p`
  width: 100%;
  color: ${colors.label};
  font-weight: 300;
  font-size: ${fontSizes.base}px;
  line-height: normal;
  margin-bottom: 10px;
`;

const ProjectLink = styled(Link)`
  color: inherit;
  font-weight: 400;
  font-size: inherit;
  line-height: inherit;
  text-decoration: underline;
  transition: all 100ms ease-out;
  margin-left: 4px;

  &:hover {
    color: ${darken(0.2, colors.label)};
    text-decoration: underline;
  }
`;

const IdeaImage = styled.img`
  width: 100%;
  margin: 0 0 2rem;
  padding: 0;
  border-radius: 8px;
  border: 1px solid ${colors.separation};
`;

const StyledIdeaBody = styled(IdeaBody)`
  margin-bottom: 20px;
`;

const StyledIdeaMap = styled(IdeaMap)`
  margin-bottom: 40px;
`;

const StyledOfficialFeedback = styled(OfficialFeedback)`
  margin-top: 70px;
`;

const StyledComments = styled(Comments)`
  margin-top: 30px;
`;

const Right = styled.div`
  flex: 2;
  position: sticky;
  top: 80px;
  align-self: flex-start;
  color: ${colors.adminTextColor};
  font-size: ${fontSizes.base}px;
  line-height: 19px;
`;

const BudgetBox = styled.div`
  margin-top: 20px;
  width: 100%;
  height: 95px;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 5px;
  position: relative;
  border-radius: 5px;
  background: ${colors.background};
  border: solid 1px ${colors.adminTextColor};
  font-size: ${fontSizes.large}px;
  font-weight: 500;
`;

const Picks = styled.div`
  margin-top: 7px;
  display: flex;
  font-size: ${fontSizes.base}px;
  align-items: center;
`;

interface State {}

interface InputProps {
  ideaId: string | null;
  closePreview: () => void;
  handleClickEdit: () => void;
}

interface DataProps {
  idea: GetIdeaChildProps;
  ideaImages: GetIdeaImagesChildProps;
  ideaFiles: GetResourceFilesChildProps;
  tenant: GetTenantChildProps;
  project: GetProjectChildProps;
}

interface Props extends InputProps, DataProps {}

export class IdeaContent extends PureComponent<Props & InjectedLocalized & InjectedIntlProps, State> {
  handleClickDelete = () => {
    const { idea, closePreview } = this.props;
    const message = this.props.intl.formatMessage(messages.deleteIdeaConfirmation);

    if (!isNilOrError(idea)) {
      if (window.confirm(message)) {
        deleteIdea(idea.id);
        closePreview();
      }
    }
  }

  render() {
    const {
      idea,
      project,
      localize,
      ideaImages,
      ideaFiles,
      tenant,
      handleClickEdit,
      intl: { formatMessage }
    } = this.props;

    if (!isNilOrError(idea)) {
      const ideaTitle = localize(idea.attributes.title_multiloc);
      const ideaImageLarge = !isNilOrError(ideaImages) && ideaImages.length > 0 ? get(ideaImages[0], 'attributes.versions.large', null) : null;
      const ideaGeoPosition = (idea.attributes.location_point_geojson || null);
      const ideaAddress = (idea.attributes.location_description || null);

      return (
        <Container>
          <Top>
            <Button
              icon="edit"
              style="text"
              textColor={colors.adminTextColor}
              onClick={handleClickEdit}
            >
              <FormattedMessage {...messages.edit}/>
            </Button>
            <Button
              icon="delete"
              style="text"
              textColor={colors.adminTextColor}
              onClick={this.handleClickDelete}
            >
              <FormattedMessage {...messages.delete}/>
            </Button>
          </Top>
          <Content>
            {!isNilOrError(project) &&
              <BelongsToProject>
                <FormattedMessage
                  {...messages.postedIn}
                  values={{
                    projectLink:
                      <ProjectLink className="e2e-project-link" to={`/projects/${project.attributes.slug}`}>
                        <T value={project.attributes.title_multiloc} />
                      </ProjectLink>
                  }}
                />
              </BelongsToProject>
            }

            <StyledIdeaTitle
              ideaId={idea.id}
              ideaTitle={ideaTitle}
            />
            <Row>
              <Left>
                {ideaImageLarge &&
                  <IdeaImage src={ideaImageLarge} alt={formatMessage(messages.imageAltText, { ideaTitle })} className="e2e-ideaImage"/>
                }
                <IdeaAuthor
                  authorId={get(idea, 'relationships.author.data.id', null)}
                  ideaCreatedAt={idea.attributes.created_at}
                  ideaId={idea.id}
                />

                <StyledIdeaBody
                  ideaId={idea.id}
                  ideaBody={localize(idea.attributes.body_multiloc)}
                />

                {ideaGeoPosition && ideaAddress &&
                  <StyledIdeaMap
                    address={ideaAddress}
                    position={ideaGeoPosition}
                    id={idea.id}
                  />
                }

                {ideaFiles && !isNilOrError(ideaFiles) &&
                  <FileAttachments files={ideaFiles} />
                }

                <StyledOfficialFeedback
                  ideaId={idea.id}
                />

                <StyledComments ideaId={idea.id} />
              </Left>
              <Right>
                <VotePreview ideaId={idea.id}/>

                {idea.attributes.budget && !isNilOrError(tenant) &&
                  <>
                    <BudgetBox>
                      <FormattedNumber
                        value={idea.attributes.budget}
                        style="currency"
                        currency={tenant.attributes.settings.core.currency}
                        minimumFractionDigits={0}
                        maximumFractionDigits={0}
                      />
                      <Picks>
                        <FormattedMessage {...messages.picks} values={{ picksNumber: idea.attributes.baskets_count }} />
                        &nbsp;
                        <InfoTooltip {...messages.basketsCountTooltip} size="small" position="top-left" />
                      </Picks>
                    </BudgetBox>
                  </>
                }

                <IdeaSettings ideaId={idea.id}/>
              </Right>
            </Row>
          </Content>
        </Container>
      );
    }
    return null;
  }
}

const Data = adopt<DataProps, InputProps>({
  idea: ({ ideaId, render }) => <GetIdea id={ideaId}>{render}</GetIdea>,
  project: ({ idea, render }) => <GetProject id={get(idea, 'relationships.project.data.id')}>{render}</GetProject>,
  ideaFiles: ({ ideaId, render }) => <GetResourceFiles resourceId={ideaId} resourceType="idea">{render}</GetResourceFiles>,
  ideaImages: ({ ideaId, render }) => <GetIdeaImages ideaId={ideaId}>{render}</GetIdeaImages>,
  tenant: <GetTenant />,
});

const IdeaContentWithHOCs = injectIntl(injectLocalize(IdeaContent));

const WrappedIdeaContent = (inputProps: InputProps) => (
  <Data {...inputProps}>
    {dataProps => <IdeaContentWithHOCs {...inputProps} {...dataProps} />}
  </Data>
);

export default WrappedIdeaContent;
