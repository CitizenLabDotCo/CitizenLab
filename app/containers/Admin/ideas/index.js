import React from 'react';
import PropTypes from 'prop-types';
import HelmetIntl from 'components/HelmetIntl';

// components
// import { FormattedMessage } from 'react-intl';

// store
import { preprocess } from 'utils';

// messages
import messages from './messages';
import { resetIdeas } from 'resources/ideas/actions';

class IdeaDashboard extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function

  componentWillUnmount() {
    // reset ideas upon page leaving, to avoid double items when going back
    this.props.resetIdeas();
  }

  render() {
    return (
      <div>
        <HelmetIntl
          title={messages.helmetTitle}
          description={messages.helmetDescription}
        />
        {this.props.children}
      </div>

    );
  }
}


IdeaDashboard.propTypes = {
  resetIdeas: PropTypes.func.isRequired,
  children: PropTypes.element,
};

export default preprocess(null, { resetIdeas })(IdeaDashboard);
