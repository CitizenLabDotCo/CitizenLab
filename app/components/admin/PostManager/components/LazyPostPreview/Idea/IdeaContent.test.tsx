import React from 'react';
import { shallow } from 'enzyme';
import { localizeProps } from 'utils/testUtils/localizeProps';
import { getDummyIntlObject } from 'utils/testUtils/mockedIntl';
import { getIdea } from 'services/__mocks__/ideas';

jest.mock('containers/IdeasShow/PostedBy', () => 'PostedBy');
jest.mock(
  'containers/IdeasShow/IdeaProposedBudget',
  () => 'IdeaProposedBudget'
);
jest.mock('components/PostShowComponents/Title', () => 'Title');
jest.mock('components/PostShowComponents/Body', () => 'Body');
jest.mock('components/PostShowComponents/DropdownMap', () => 'DropdownMap');
jest.mock(
  'components/PostShowComponents/OfficialFeedback',
  () => 'OfficialFeedback'
);
jest.mock('components/PostShowComponents/Comments', () => 'Comments');
jest.mock('components/UI/FileAttachments', () => 'FileAttachments');
jest.mock('components/UI/Button', () => 'Button');
jest.mock('cl2-component-library', () => ({ IconTooltip: 'IconTooltip' }));
jest.mock('../PostPreview', () => ({
  Top: 'Top',
  Content: 'Content',
  Container: 'Container',
}));
jest.mock('./FeedbackSettings', () => 'FeedbackSettings');
jest.mock('./VotePreview', () => 'VotePreview');
jest.mock('utils/cl-router/Link', () => 'Link');
jest.mock('components/T', () => 'T');
jest.mock('services/ideas');
jest.mock('resources/GetResourceFiles', () => 'GetResourceFiles');
jest.mock('resources/GetIdea', () => 'GetIdea');
jest.mock('resources/GetIdeaImages', () => 'GetIdeaImages');
jest.mock('resources/GetTenant', () => 'GetTenant');
jest.mock('resources/GetProject', () => 'GetProject');
jest.mock('resources/GetPermission', () => 'GetPermission');
jest.mock('resources/GetLocale', () => 'GetLocale');
jest.mock('utils/cl-intl');
jest.mock('utils/styleUtils', () => ({
  colors: {
    label: '#000',
    separation: '#000',
    adminTextColor: '#000',
  },
  fontSizes: {
    base: 16,
  },
  media: {
    smallerThanMinTablet: (x) => x,
  },
}));

import { IdeaContent } from './IdeaContent';

describe('<IdeaContent />', () => {
  let closePreview: jest.Mock;
  let handleClickEdit: jest.Mock;

  beforeEach(() => {
    closePreview = jest.fn();
    handleClickEdit = jest.fn();
  });

  it('renders correctly when ideaId is not defined', () => {
    const intl = getDummyIntlObject();
    const wrapper = shallow(
      <IdeaContent
        ideaId={null}
        closePreview={closePreview}
        handleClickEdit={handleClickEdit}
        idea={null}
        ideaImages={null}
        ideaFiles={null}
        tenant={null}
        project={null}
        locale={null}
        postOfficialFeedbackPermission={undefined}
        intl={intl}
        {...localizeProps}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly when an idea is provided', () => {
    const ideaId = 'myIdeasiD';
    const idea = getIdea(ideaId);
    const intl = getDummyIntlObject();
    const wrapper = shallow(
      <IdeaContent
        ideaId={ideaId}
        closePreview={closePreview}
        handleClickEdit={handleClickEdit}
        idea={idea}
        ideaImages={null}
        ideaFiles={null}
        tenant={null}
        project={null}
        locale={null}
        postOfficialFeedbackPermission={undefined}
        intl={intl}
        {...localizeProps}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });
});
