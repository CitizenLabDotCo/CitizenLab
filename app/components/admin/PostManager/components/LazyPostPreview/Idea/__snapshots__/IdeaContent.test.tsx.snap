// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<IdeaContent /> renders correctly when an idea is provided 1`] = `
<Container>
  <Top>
    <Button
      buttonStyle="text"
      icon="edit"
      onClick={[MockFunction]}
    >
      <FormattedMessage
        defaultMessage="Edit"
        id="app.components.admin.PostManager.PostPreview.edit"
      />
    </Button>
    <Button
      buttonStyle="text"
      icon="delete"
      onClick={[Function]}
    >
      <FormattedMessage
        defaultMessage="Delete"
        id="app.components.admin.PostManager.PostPreview.delete"
      />
    </Button>
  </Top>
  <Content>
    <IdeaContent__StyledTitle
      postId="myIdeasiD"
      postType="idea"
      title="Doloribus quam molestiae ut."
    />
    <IdeaContent__Row>
      <IdeaContent__Left>
        <IdeaAuthor
          authorId="8ea3a61c-e1bd-4756-a54d-c092553abccd"
          ideaId="myIdeasiD"
          ideaPublishedAt="2019-03-19T00: 00: 00.000Z"
        />
        <IdeaContent__BodySectionTitle>
          <FormattedMessage
            defaultMessage="Proposed budget"
            id="app.containers.IdeasShow.proposedBudgetTitle"
          />
        </IdeaContent__BodySectionTitle>
        <Component
          proposedBudget={45}
        />
        <IdeaContent__BodySectionTitle>
          <FormattedMessage
            defaultMessage="Description"
            id="app.containers.IdeasShow.bodyTitle"
          />
        </IdeaContent__BodySectionTitle>
        <IdeaContent__StyledBody
          body="<p>Ea ea aut. Quasi corrupti iste. Delectus eum voluptates.</p><p>Est harum voluptates. Et fugit enim. Suscipit inventore ullam.</p><p>Eos mollitia omnis. Laudantium porro perferendis. Minus in aut.</p>"
          postId="myIdeasiD"
          postType="idea"
        />
        <IdeaContent__StyledOfficialFeedback
          permissionToPost={true}
          postId="myIdeasiD"
          postType="idea"
        />
        <IdeaContent__StyledComments
          postId="myIdeasiD"
          postType="idea"
        />
      </IdeaContent__Left>
      <IdeaContent__Right>
        <VotePreview
          ideaId="myIdeasiD"
        />
        <FeedbackSettings
          ideaId="myIdeasiD"
        />
      </IdeaContent__Right>
    </IdeaContent__Row>
  </Content>
</Container>
`;

exports[`<IdeaContent /> renders correctly when ideaId is not defined 1`] = `""`;
